### PR TITLE
Ed25519 test cases

### PR DESCRIPTION
**Pull request by sa-kib**
Add test cases to generate ED25519 key, sign/verify for ED25519, ED25519ctx and ED25519ph.
Companion PR for Ed25519 OP-TEE implementation: [#5486](https://github.com/OP-TEE/optee_os/pull/5486)

### CODE REVIEW

File: host/xtest/regression_4000_data.h [hunk 1]
```diff
@@ -8650,4 +8650,94 @@ static const uint8_t x25519_shared_secret[] = {
 	0x76, 0xf0, 0x9b, 0x3c, 0x1e, 0x16, 0x17, 0x42
 };
 
+
```
Comment (jenswi-linaro) [inline host/xtest/regression_4000_data.h:8653]: Please add a comment describing where you got these test vectors from.

File: host/xtest/regression_4000.c [hunk 2]
```diff
@@ -5766,4 +5864,135 @@ static void xtest_tee_test_4015(ADBG_Case_t *c)
 }
 ADBG_CASE_DEFINE(regression, 4015, xtest_tee_test_4015,
 		"Test TEE Internal API Derive key X25519");
+
+static void xtest_tee_test_4016_ed25519(ADBG_Case_t *c)
+{
+	TEEC_Session session = { };
+	TEE_OperationHandle op = TEE_HANDLE_NULL;
+	TEE_ObjectHandle key_handle = TEE_HANDLE_NULL;
+
```
Comment (jenswi-linaro) [inline host/xtest/regression_4000.c:5873]: Empty line not needed between the declarations of stack variables.

File: host/xtest/regression_4000.c [hunk 3]
```diff
@@ -5766,4 +5864,135 @@ static void xtest_tee_test_4015(ADBG_Case_t *c)
...
+	uint8_t out[64] = { };
+	uint32_t out_size = sizeof(out);
+
+	size_t n = 0;
+	uint32_t ret_orig = 0;
+	size_t max_key_size = 0;
+
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+			xtest_teec_open_session(&session, &crypt_user_ta_uuid,
+			    NULL, &ret_orig)))
```
Comment (jenswi-linaro) [inline host/xtest/regression_4000.c:5888]: Please align with the `(` on the line above. Same below where possible.

Comment (sa-kib) [reply host/xtest/regression_4000.c]: done in [21de224](https://github.com/OP-TEE/optee_test/pull/610/commits/21de2241a42fc9ab579909deb55e0653a6faf5bf)

File: host/xtest/regression_4000_data.h (hunk 1 above)
Comment (sa-kib) [reply host/xtest/regression_4000_data.h]: done in [21de224](https://github.com/OP-TEE/optee_test/pull/610/commits/21de2241a42fc9ab579909deb55e0653a6faf5bf)

File: host/xtest/regression_4000.c [hunk 4]
```diff
@@ -5766,4 +5864,129 @@ static void xtest_tee_test_4015(ADBG_Case_t *c)
...
+				goto out;
+
+			if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+					ta_crypt_cmd_asymmetric_sign(c,
+						&session, op,
+						attrs, num_attrs, tv->ptx,
+						tv->ptx_len, out, &out_size)))
+				goto out;
+
+			ADBG_EXPECT_COMPARE_UNSIGNED(c,
```
Comment (jenswi-linaro) [inline host/xtest/regression_4000.c:5950]: Please use `ADBG_EXPECT_BUFFER()` instead.

**Review by jenswi-linaro — COMMENTED** [review]
With my comment addressed:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: host/xtest/regression_4000.c (hunk 4 above)
Comment (sa-kib) [reply host/xtest/regression_4000.c]: done

File: host/xtest/regression_4000.c [hunk 5]
```diff
@@ -2722,6 +2722,15 @@ struct xtest_ac_case {
...
 		} ecc;
+	struct {
+		const uint8_t *private;
+		size_t private_len;
+		const uint8_t *public;
+		size_t public_len;
+		const uint8_t flag;
+		const uint8_t *context;
+		size_t context_len;
+	} eddsa;
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:2733]: fix indentation

File: host/xtest/regression_4000.c [hunk 6]
```diff
@@ -5766,4 +5864,128 @@ static void xtest_tee_test_4015(ADBG_Case_t *c)
...
+
+		if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+				ta_crypt_cmd_free_operation(c, &session, op)))
+			goto out;
+	}
+out:
+	TEEC_CloseSession(&session);
+}
+ADBG_CASE_DEFINE(regression, 4016_ed25519, xtest_tee_test_4016_ed25519,
+"Test TEE Internal API ED25519 sign/verify");
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:5989]: fix indenation

```patch
 ADBG_CASE_DEFINE(regression, 4016_ed25519, xtest_tee_test_4016_ed25519,
-"Test TEE Internal API ED25519 sign/verify");
+		  "Test TEE Internal API ED25519 sign/verify");

File: host/xtest/regression_4000.c [hunk 7]
```diff
@@ -5766,4 +5864,128 @@ static void xtest_tee_test_4015(ADBG_Case_t *c)
...
+		xtest_add_attr(&num_key_attrs, key_attrs,
+			       TEE_ATTR_ED25519_PUBLIC_VALUE,
+			       tv->params.eddsa.public,
+			       tv->params.eddsa.public_len);
+
+		if (tv->params.eddsa.flag == 1)
+			xtest_add_attr(&num_attrs, attrs,
+				       TEE_ATTR_EDDSA_PREHASH, NULL, 0);
+
+
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:5918]: remove one of the 2 empty lines

**Review by etienne-lms — CHANGES_REQUESTED** [review]
There are minor indentation issues.
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` once addressed.

File: host/xtest/regression_4000.c [hunk 8]
```diff
@@ -4747,6 +4813,38 @@ static void xtest_tee_test_4007_x25519(ADBG_Case_t *c)
...
+		goto out;
+	}
+
+	Do_ADBG_BeginSubCase(c, "Generate Ed25519 key");
+
+	if (!ADBG_EXPECT_TRUE(c,
+			generate_and_test_key(c, &session,
+					      TEE_TYPE_ED25519_KEYPAIR, 0, 256,
+					      NULL, 0)))
+		return;
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:4839]: shouold remove instruction `return` here:

```c
	Do_ADBG_BeginSubCase(c, "Generate Ed25519 key");

	ADBG_EXPECT_TRUE(c, generate_and_test_key(c, &session,
						   TEE_TYPE_ED25519_KEYPAIR,
						   0, 256, NULL, 0));

	Do_ADBG_EndSubCase(c, "Generate Ed25519 key");
out:
	TEEC_CloseSession(&session);

**Comment by jforissier** [comment]
> There is a problem with the key generation test
> 
> ```
>   2067:  * regression_4007_ed25519 Test TEE Internal API Generate ed25519 key
>   2068:  o regression_4007_ed25519.1 Generate Ed25519 key
>   2069:  regression_4000.c:4350: ta_crypt_cmd_generate_key(c, s, key, key_size, params, param_count) has an unexpected value: 0xffff0006 = TEEC_ERROR_BAD_PARAMETERS, expected 0x0 = TEEC_SUCCESS
>   2070:  regression_4000.c:4835: generate_and_test_key(c, &session, 0xA1000043, 0, 256, ((void *)0), 0) has an unexpected value: 0x0 = false, expected 0x1 = true
>   2071:    regression_4007_ed25519.1 FAILED
>   2072:    regression_4007_ed25519 FAILED
> ```

https://github.com/OP-TEE/optee_os/pull/5569

**Comment by jforissier** [comment]
Yet another problem with the key generation test, this time hopefully fixed by https://github.com/OP-TEE/optee_os/pull/5573.

File: host/xtest/regression_4000.c (hunk 5 above)
Comment (sa-kib) [reply host/xtest/regression_4000.c]: fixed

File: host/xtest/regression_4000.c (hunk 6 above)
Comment (sa-kib) [reply host/xtest/regression_4000.c]: fixed

File: host/xtest/regression_4000.c (hunk 7 above)
Comment (sa-kib) [reply host/xtest/regression_4000.c]: removed

File: host/xtest/regression_4000.c (hunk 8 above)
Comment (sa-kib) [reply host/xtest/regression_4000.c]: done, thank you!

**Comment by sa-kib** [comment]
> @sa-kib would you mind addressing @etienne-lms's comments and applying his Acked-by? Thanks!

patch is updated now, thank you!

**Comment by jforissier** [comment]
CI is now happy, thanks all.

**Pull request merged**